pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/AuthenticationProviderTokenOIDC.java

```diff
@@ -0,0 +1,361 @@
+
+    static final String TOKEN = "token";
+
+    private static final Counter expiredTokenMetrics = Counter.build()
+            .name("pulsar_expired_token_count")
+            .help("Pulsar expired token")
+            .register();
+
+    private static final Histogram expiringTokenMinutesMetrics = Histogram.build()
+            .name("pulsar_expiring_token_minutes")
+            .help("The remaining time of expiring token in minutes")
+            .buckets(5, 10, 60, 240)
+            .register();
+
+    private String roleClaim;
+    private String audienceClaim;
+    private String audience;
+    private JwkProvider provider;
+    private String issuerUrl;
+
+    // config keys
+
+    private String confTokenAuthClaimSettingName;
+    private String confTokenAudienceClaimSettingName;
+    private String confTokenAudienceSettingName;
+    private String confJWkUrlSettingName;
+
+
+
+    @Override
+    public void close() throws IOException {
+        // noop
+    }
+
+    @VisibleForTesting
+    public static void resetMetrics() {
+        expiredTokenMetrics.clear();
+        expiringTokenMinutesMetrics.clear();
+    }
+
+    @Override
+    public void initialize(ServiceConfiguration config) throws IOException, IllegalArgumentException {
+        this.confTokenAuthClaimSettingName = CONF_TOKEN_AUTH_CLAIM;
+        this.confTokenAudienceClaimSettingName = CONF_TOKEN_AUDIENCE_CLAIM;
+        this.confTokenAudienceSettingName = CONF_TOKEN_AUDIENCE;
+        this.confJWkUrlSettingName = CONF_ISSUER_URL;
+
+        // we need to fetch the algorithm before we fetch the key
+        this.roleClaim = getTokenRoleClaim(config);
+        this.audienceClaim = getTokenAudienceClaim(config);
+        this.audience = getTokenAudience(config);
+        this.issuerUrl = getIssuerUrl(config);
+
+        URL url = new URL(this.issuerUrl+"/.well-known/openid-configuration");
+        if(!url.getProtocol().equals("https")){
+            throw new MalformedURLException("protocol needs to be https");
+        }
+        //extracting the jwks_uri
+        JsonObject json = new Gson().fromJson(IOUtils.toString(url, UTF_8), JsonObject.class);
+        this.provider = new UrlJwkProvider(new URL(json.get("jwks_uri").getAsString()));
+
+
+
+        if (audienceClaim != null && audience == null ) {
+            throw new IllegalArgumentException("Token Audience Claim [" + audienceClaim
+                    + "] configured, but Audience stands for this broker not.");
+        }
+    }
+
+    @Override
+    public String getAuthMethodName() {
+        return TOKEN;
+    }
+
+    @Override
+    public String authenticate(AuthenticationDataSource authData) throws AuthenticationException {
+        try {
+            // Get Token
+            String token;
+            String role;
+            token = getToken(authData);
+
+            // Parse Token by validating
+
+
+            role = getPrincipal(authenticateToken(token));
```

Comment:
Consider storing the result of `authenticateToken` into a local variable before invoking `getPrincipal`.  Anticipate that other claims might be extracted from the token in future.